test/upgrade/types.go

```diff
@@ -81,45 +81,29 @@ type Context struct {
 }
 
 // BackgroundContext is a upgrade test execution context that will be passed
-// down to each handler of BackgroundOperation. It contains a StopEvent channel
-// which end user should use to obtain a testing.T for error reporting. Until
-// StopEvent is sent user may use zap.SugaredLogger to log state of execution if
-// necessary. The logs are stored in a threadSafeBuffer and flushed to the test
-// output when the test fails.
+// down to each handler of BackgroundOperation. It contains a stop event channel.
+// Until stop event is sent user may use zap.SugaredLogger to log state of execution if
+// necessary.
 type BackgroundContext struct {
-	Log       *zap.SugaredLogger
-	Stop      <-chan StopEvent
-	logBuffer *threadSafeBuffer
-}
-
-// StopEvent represents an event that is to be received by background operation
-// to indicate that is should stop it's operations and validate results using
-// passed T. User should use Finished channel to signalize upgrade suite that
-// all stop & verify operations are finished and it is safe to end tests.
-type StopEvent struct {
-	T        *testing.T
-	Finished chan<- struct{}
-	name     string
-	logger   *zap.SugaredLogger
+	T    *testing.T
+	Log  *zap.SugaredLogger
```

Comment:
Should each operation get a logger that's backed by the testing.T - ideally it's not a global test object but the one scoped under `t.Run(name, func(t *testing.T))`?

https://pkg.go.dev/go.uber.org/zap/zaptest#NewLogger

Then I don't think you have to worry about printing on failures.

Reply:
Great point! This was actually a bug in my code because it was not printing the logs even on failure.
Pushing another commit now that fixes it. I've tested this new change in [serving run](https://storage.googleapis.com/knative-prow/pr-logs/pull/knative_serving/13568/upgrade-tests_serving_main/1611279118299566080/build-log.txt) and it now properly logs stuff like `INFO	3) 🔄 Starting continual tests. 3 tests are registered.` and similar.